python/cudf/cudf/core/index.py

```diff
@@ -58,11 +58,33 @@ def _to_frame(this_index, index=True, name=None):
 
 
 class Index(Frame, Serializable):
-    def __new__(cls, data=None, dtype=None, name=None, **kwargs):
+    def __new__(
+        cls,
+        data=None,
+        dtype=None,
+        copy=False,
+        name=None,
+        tupleize_cols=True,
+        **kwargs,
+    ):
+        if tupleize_cols is not True:
+            raise NotImplementedError(
+                "tupleize_cols != True is not yet supported"
+            )
 
+        if copy is True:
+            data = data.copy()
```

Comment:
What if `data` is an object that doesn't have a `.copy()` method? I.e. a tuple

Reply:
Also if `data` is a host object this will unnecessarily make an extra copy in host memory which could be expensive.

Reply:
hmm, I agree. Should we in that case check if it is a device object and then only use copy. Mostly it would be helpful in case of cudf.Column.

Reply:
It should be `__cuda_array_interface__` but just checking for that attribute doesn't work well enough here, because for example if we have a float with `NaN` values, by default we copy the column to replace them with nulls.

I'd suggest if `copy=True` we just always do `as_column(data).copy()` so at least it's a device copy and not potentially a host copy.

Reply:
Bur doing as_column(data).copy() will be creating two sets of device memory if data is host, even though memory may not last long, but there can be a memory shoot at that point.

Reply:
Yes it would, but only in the case where they explicitly pass `copy=True` which is already a potentially slow path. Otherwise we'd have to know when we're dealing with a host vs GPU object and only do `as_column(data).copy()` on the GPU object.